server/sonar-server/src/main/java/org/sonar/server/computation/component/ComponentImpl.java

```diff
@@ -0,0 +1,142 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.computation.component;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.sonar.api.measures.Metric;
+import org.sonar.batch.protocol.Constants;
+import org.sonar.batch.protocol.output.BatchReport;
+import org.sonar.core.measure.db.MeasureDto;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.server.computation.ComputationContext;
+import org.sonar.server.computation.event.Event;
+import org.sonar.server.computation.event.EventRepository;
+import org.sonar.server.computation.measure.MeasureRepository;
+import org.sonar.server.db.DbClient;
+
+import static com.google.common.base.Predicates.notNull;
+import static com.google.common.collect.ImmutableList.copyOf;
+import static com.google.common.collect.Iterables.filter;
+import static java.util.Objects.requireNonNull;
+
+public class ComponentImpl implements Component {
+  private final ComputationContext context;
+  private final Type type;
+  private final BatchReport.Component component;
+  private final List<Component> children;
+  private final EventRepository eventRepository = new SetEventRepository();
+
+  public ComponentImpl(ComputationContext context, BatchReport.Component component, @Nullable Iterable<Component> children) {
+    this.context = context;
+    this.component = component;
+    this.type = convertType(component.getType());
+    this.children = children == null ? Collections.<Component>emptyList() : copyOf(filter(children, notNull()));
+  }
+
+  private static Type convertType(Constants.ComponentType type) {
+    switch (type) {
+      case PROJECT:
+        return Type.PROJECT;
+      case MODULE:
+        return Type.MODULE;
+      case DIRECTORY:
+        return Type.DIRECTORY;
+      case FILE:
+        return Type.FILE;
+      default:
+        throw new IllegalArgumentException("Unsupported Constants.ComponentType value " + type);
+    }
+  }
+
+  @Override
+  public Type getType() {
+    return type;
+  }
+
+  @Override
+  public int getRef() {
+    return component.getRef();
+  }
+
+  @Override
+  public List<Component> getChildren() {
+    return children;
+  }
+
+  @Override
+  public org.sonar.server.computation.context.ComputationContext getContext() {
+    return context;
+  }
+
+  @Override
+  public EventRepository getEventRepository() {
+    return eventRepository;
+  }
+
+  @Override
+  public MeasureRepository getMeasureRepository() {
```

Comment:
I think this repository is missing something : how will we be able to get computed measures by the CE ?
And reading measures from DB for each component seems not the best option, but we'll change it in the next refactoring